### ISSUE TITLE
Type mismatch in MPI and netCDF when using `#include` with GCC 10

### ISSUE THREAD

**Issue by marshallward** [NOAA-GFDL/FMS#561]
**Describe the bug**
When building FMS with GCC 10, several errors arise due to "type mismatch":
```
../src/mpp/include/mpp_comm_mpi.inc:447:44:

  447 |       if( mpp_npes().GT.1 ) call MPI_BCAST( data, length*size(data(:)), MPI_CHARACTER, from_rank, peset(n)%id, error )
      |                                            1
Error: Type mismatch between actual argument at (1) and actual argument at (2) (CHARACTER(0)/LOGICAL(4)).
```
The error above is raised because `MPI_Bcast` is called with several different types (integer, real, logical, etc) and the interface for `MPI_Bcast` is unavailable (explained below).  In the absence of such an interface, the compiler assumes that his function will be defined later.  If the function receives different sets of arguments, then this is is a potential error.  (I believe the rules are clarified in C10.6 of the Standard, but they're very detailed and hard for me to follow).

This is not restricted to `MPI_Bcast`; this is happening for just about every MPI and netCDF function which relies on an interface.

The interface is unavailable because most builds are still using the older `#include` headers (see #30), which do not contain interfaces.  Until link time, these MPI and netCDF calls are handled as generic undefined functions.

The error is now occurring because GCC 10 has stricter requirements for such type mismatches.

This can be resolved by replacing the `#include` with `use` statements to the MPI and netCDF modules.

I am willing to do this work, but have two questions:

* Remove explicit references to `sgi_mipspro` when setting up the MPI module.
  
  It is still feasible to toggle between `use mpi` and `#include <mpif.h>` if need be, but it can be done via autoconf rather than a platform-specific header if needed.  But I wonder if there is any library left which still requires this header?

  There is also the matter that using `#include` here is not compliant with the MPI standard.

* Replace `nf_*` style netCDF functions with `nf90_*` style, alongside the introduction of `use netcdf`.
  
  This will be a bit more work, but is mostly just name substitution.  However, it potentially clashes with ongoing IO work.

**To Reproduce**
Build FMS using GCC 10 gfortran compiler.

**Expected behavior**
The library should build as expected.

**System Environment**
Describe the system environment, include:
- OS:
```
$ uname -a
Linux guadosalam 5.7.12-arch1-1 #1 SMP PREEMPT Fri, 31 Jul 2020 17:38:22 +0000 x86_64 GNU/Linux
```
- Compiler(s): Type and version [e.g. Intel 19.1]
```
$ gfortran --version
GNU Fortran (GCC) 10.1.0
```
- MPI type, and version (e.g. MPICH, Cray MPI, openMPI)
```
ompi_info --version
Open MPI v4.0.4
```
- netCDF Version: For both C and Fortran
```
$ nc-config --version
netCDF 4.7.4
```
- Configure options: Any additional flags, or macros passed to configure

Largely standard (still mkmf-based so ping me if more info is needed).

**Additional context**
N/A

**Comment by marshallward** [comment]
I forgot to add that the `-fallow-argument-mismatch` flag will turn these errors into warnings, which I am using locally, but the underlying problem appears to be something that can be fixed.

**Comment by bensonr** [comment]
@marshallward  Thank you for opening the issue.  We are currently using the workaround on our internal machine with gcc-10.2.0.  I am in the process of doing just what you mentioned in above and will take responsibility for pushing the fixes.

**Comment by thomas-robinson** [comment]
@bensonr This is solved, correct?  I will close this issue, and if it needs more work, you can reopen it.  Otherwise no response in necessary.

**Comment by bensonr** [comment]
Yes this issue was referenced as one of the issues address and fixed in PR #580
